man/attester.1.md
% attester(1) CHARRA

# NAME

**attester**(1) - Perform remote attestation using CHARRA.

# SYNOPSIS

**attester** [*OPTIONS*]

# DESCRIPTION

**attester**(1) - The primary role of the attester in CHARRA is to act as the entity being attested. Upon receiving a challenge from the verifier, the attester gathers the requested measurements by performing a TPM quote operation. This quote includes the selected Platform Configuration Register (PCR) values, which reflect the current state of the system. The attester then sends this signed evidence back to the verifier for evaluation.

# OPTIONS

  * **-h**, **\--help**:

    Prints usage information.

  * **-c**, **\--config=PATH**:

    Loads attester config from a file.

  * **-v**, **\--verbose**:

    Set CHARRA and CoAP log-level to DEBUG.

  * **-l**, **\--log-level=LEVEL**:

    Set CHARRA log-level to LEVEL. Available are: TRACE, DEBUG, INFO, WARN, ERROR, FATAL. Default is INFO.

  * **\--coap-log-level=LEVEL**:

    Set CoAP log-level to LEVEL. Available are: DEBUG, INFO, NOTICE, WARNING, ERR, CRIT, ALERT, EMERG, CIPHERS. Default is INFO.

  * **-k**, **\--attestation-key=FORMAT:VALUE**:

    Specifies the path to the attestation key. Available are: context, handle.

  * **\--port=PORT**:

    Define a custom port.

  * **\--pcr-log=FORMAT:FILE**:

    Specifies the path to the PCR log file. Available formats are: ima, tcg-boot.

  * **\--scheme=SCHEME**:

    Specifies the signature scheme used for the TPM2 quote.

  * **-g**, **\--hash-algorithm=ALGORITHM**

    Specifies the hash algorithm used for the TPM2 quote.

# DTLS-PSK Options:

* **-p**, **\--psk**:

    Enable DTLS protocol with PSK. By default, the key 'CHARRA DTLS Key' and hint 'CHARRA Attester' are used.

* **\--psk-key=KEY**:

    Use KEY as pre-shared key for DTLS. Implicitly enables DTLS-PSK.

* **\--psk-hint=HINT**:

    Use HINT as hint for DTLS. Implicitly enables DTLS-PSK.

# DTLS-RPK Options:

CHARRA includes default keys in the 'keys' folder, but these are only intended for testing. They MUST be changed in actual production environments!

*  **-r**, **\--rpk**:

    Enable DTLS-RPK (raw public keys) protocol. The protocol is intended for scenarios in which public keys of either attester or verifier or both of them are pre-shared.

*  **\--rpk-private-key=PATH**:

    Specify the path of the private key used for RPK. Currently, only supports DER (ASN.1) format.

    By default, 'keys/attester.der' is used. Implicitly enables DTLS-RPK.

*  **\--rpk-public-key=PATH**:

    Specify the path of the public key used for RPK. Currently, only supports DER (ASN.1) format.

    By default, 'keys/attester.pub.der' is used. Implicitly enables DTLS-RPK.

*  **\--rpk-peer-public-key=PATH**:

    Specify the path of the reference public key of the peer, used for RPK. Currently, only supports DER (ASN.1) format.

    By default, 'keys/verifier.pub.der' is used. Implicitly enables DTLS-RPK.

*  **\--rpk-verify-peer=[0,1]**:

    Specify whether the peer's public key shall be checked against the reference public key. 0 means no check, 1 means check. By default, the check is performed.

    WARNING: Disabling the verification means that connections from any peer will be accepted. This is primarily intended for the verifier, which may not have
    the public keys of all attesters and does an identity check with the attestation response. Implicitly enables DTLS-RPK.

# TCTI

To configure TCTI for the TPM, set the 'CHARRA_TCTI' environment variable accordingly.

# CONFIGURATION FILES

CHARRA's attester can be configured using a YAML configuration file. By default, the attester automatically loads **/etc/charra/attester-config.yaml** if it exists. Other specified config files and command-line arguments override values from a config file unless **lock-config** is set to **true**.

## Configuration Structure

Below is a description of the configuration structure and its available fields:

### Top-Level: charra-attester

| Field              | Type              | Description                                                          |
| ------------------ | ----------------- | -------------------------------------------------------------------- |
| lock-config        | Boolean           | If true, disables overriding this config file                        |
| listen-ip          | String            | IP address the attester binds to (e.g., "0.0.0.0")                   |
| listen-port        | Integer           | Port number the attester listens on (default: 5683)                  |
| log-level          | String            | Log level for CHARRA. One of: TRACE, DEBUG, INFO, WARN, ERROR, FATAL |

### charra-attester.coap

| Field              | Type    | Description                                                        |
| ------------------ | ------- | ------------------------------------------------------------------ |
| log-level          | String  | Log level for CoAP. Values: EMERG, ALERT, CRIT, ERR, WARNING, etc. |

### charra-attester.coap.udp.dtls-psk

| Field  | Type    | Description                                          |
| ------ | ------- | ---------------------------------------------------- |
| enable | Boolean | Enables DTLS with pre-shared keys                    |
| hint   | String  | Identity hint sent to peer (used in PSK negotiation) |
| key    | String  | The pre-shared key used for DTLS-PSK                 |


### charra-attester.coap.udp.dtls-rpk

| Field                                 | Type         | Description                                                           |
| ------------------------------------- | ------------ | --------------------------------------------------------------------- |
| enable                                | Boolean      | Enables DTLS with raw public keys (RPK)                               |
| private-key-path                      | String       | Path to the attester's private key (DER format)                       |
| public-key-path                       | String       | Path to the attester's public key (DER format)                        |
| verify-peer-public-key                | Boolean      | Whether to verify the verifier's public key against a known reference |
| peer-public-key-path                  | String       | Path to the verifier’s public key (DER format)                        |

### charra-attester.attestation

| Field                                                    | Type            | Description                                                        |
| ---------------------------------------------------      | --------------- | ------------------------------------------------------------------ |
| tpm-quote-signature-scheme                               | String          | Signature scheme to use for TPM quote (e.g. rsapss).               |
| tpm-quote-signature-hash-algorithm                       | String          | Hash algorithm to use for TPM quote signature (e.g. sha256).       |

### charra-attester.attestation.key

| Field       | Type   | Description                                                                    |
| ----------- | ------ | ------------------------------------------------------------------------------ |
| format      | String | Type of key reference: context or handle                                       |
| path        | String | Path to the key context file or a TPM handle used for TPM attestation key (AK) |

### charra-attester.attestation.pcr-log

| Field    | Type   | Description                 |
| -------- | ------ | --------------------------- |
| ima      | String | Path to IMA logs            |
| tcg-boot | String | Path to TCG boot log        |


CHANGELOG.md
<!--
################################################################################
# Copyright 2023, Fraunhofer Institute for Secure Information Technology SIT.  #
# All rights reserved.                                                         #
# ---------------------------------------------------------------------------- #
# Author:        Michael Eckel <michael.eckel@sit.fraunhofer.de>               #
# Date Modified: 2023-11-30T13:37:42+02:00                                     #
# Date Created:  2019-06-26T09:23:15+02:00                                     #
################################################################################
-->

# Changelog

The following lists the changes that CHARRA received over time.

## Changelog 2024-10-15

* Added Rust toolchain for the container user (adjustments to `Dockerfile`)

* Rust example for generating a random number with the TPM 2.0 using the `tss-esapi` crate

## Changelog 2024-06-12

* Adjustments to `cli_util`

  * Splitted the logic into `cli_util`, `cli_util_attester`, and `cli-util_verifier`

  * Prefixed `DTLS-RPK` and `DTLS_PSK` options with rpk/psk

  * Fixed undefined behavior while parsing command line error when using option `-h`

* Adjustments to `--attestation-key`

  * An attestation key can now be loaded with a TPM handle

  * `--attestation-key` will now require a FORMAT. Available are: `context` and `handle`

* `--pcr-selection` will now require a list of PCR banks as with the `tpm2-tools`

  * There is currently only support for `sha256` (other PCR banks are ignored)

* Adjustments to `verifier`

  * Added option `-g/--hash-algorithm` to set the hash-algorithm which was used to digest the tpm quote

  * The verifier is now able to verify signatures with different rsa signing schemes:

    - `TPM2_ALG_RSASSA`
    - `TPM2_ALG_RSAPSS`

## Changelog 2024-04-03

* Implemented CHARRA remote attestation using the TAP information model

  * New data transfer objects

  * Removed event log (the functionality to request logs will be added in one of the next versions)

  * reference: https://trustedcomputinggroup.org/wp-content/uploads/TNC_TAP_Information_Model_v1.00_r0.36-FINAL.pdf

* Improved interoperability with `tpm2-tools`

  * Added bash script to generate an attestation key using `tpm2-tools`

  * Attester and verifier can deserialize tpm key files generated by `tpm2-tools`

* Adjustments to `Verifier`

  * `--attestation-public-key` is a new mandatory cli argument

* Adjustments to `Attester`

  * `--attestation-key` is a new mandatory cli argument

## Changelog 2024-03-05

* Adjustments to `charra_rim_mgr` and `parser_util`:

  * Updated parser to parse the PCR reference file in a YAML format.

  * Implemented stricter parsing rules

  * Fixed minor bugs

* Adjustments to `Verifier`

  * `-f` or `--pcr-file` argument is mandatory to run the verifier

  * The specified PCR reference file has to be in this format: `FORMAT:PATH`

  * The following formats are currently available: `yaml`

* Replaced `reference-pcrs.txt` with `reference-pcrs.yml`

## Changelog 2024-02-20

* Improved `charra_marshaling` with QCBOR spiffy decoding

* Removed `cbor_util`

## Changelog 2023-11-30

* Adjustments to `Dockerfile`:

  * Updated dependencies to latest stable versions

  * Created necessary systems folders for the tpm2-tss Feature API (FAPI)

  * Added tpm2-pytss

  * Disabled TSS2 log messages that were producing (unnecessary and) confusing error messages and warnings

  * Added Bash configuration including aliases and a history (saves time typing the commands to build and run CHARRA by simply using the up/down arrow keys to cycle through the commands)

* Added TPM2 code examples under `~/code-examples`, showcasing the use of the TSS2 ESAPI and FAPI with tpm2-tss (C programming language) and tpm2-pytss (Python programming language)

* Added a script in `~/./test-charra-and-tpm2-tss.sh` for compiling and running CHARRA as well as all code examples.

## Changelog 2023-05-20

* Fixed `Makefile`

* Fixed Docker Compose

* Externalized building and installation instructions into [`INSTALL.md`](INSTALL.md) and the changelog into [`CHANGELOG.md`](CHANGELOG.md).

* Minor adjustments

## Changelog 2021-03-17

* Dynamic memory allocation for QCBOR encoded data using *malloc()*.
  Thanks, @laurencelundblade.

* Fixed some bugs

* Introduced macros for `free()`'ing heap data.

## Changelog 2021-03-16 (v2)

* Added random nonce generation with mbed TLS in Verifier.
  Made it configurable whether to use the TPM or mbed TLS to generate the nonce.

* Added media type CBOR to attestation request in Verifier.
  Credits go to @mrdeep1.

## Changelog 2021-03-16 (v1)

* Updated `README.md` to include building *tpm2software/tpm2-tss* Docker image which CHARRA uses as a basis
  Reason: recently, the official *tpm2software/tpm2-tss* Docker images were removed from [Docker Hub](https://hub.docker.com/r/tpm2software/tpm2-tss)

* Added Docker Compose file and description on how to use it to `README.md`

* Added `.editorconfig` file

* Using most recent stable versions of *tpm2-tss* and *tpm2-tools*

* Added compressed CHARRA SVG logo (`*.svgz`). See [charra-logo.svgz](./charra-logo.svgz)

## Changelog 2021-03-10

* Added support for CoAP large/block-wise data transfers, utilizing latest features of [libcoap](https://github.com/obgm/libcoap).
  This enables CHARRA to send and receive data of arbitrary size.
  Many thanks to @mrdeep1 for developing and fixing block-wise transfers in *libcoap*!

* Console output/logging can be entirely disabled with the `disable-log` Make switch.
  Colored logging can be disabled with the `disable-log-color` Make switch.
  This allows CHARRA to be used in embedded systems.
  Example:

      make disable-log=1
      make disable-log-color=1

* For debugging purposes a Make flag `address-sanitizer` was introduced. Example:

      make address-sanitizer=1

* For TPM operations a custom TCTI module can be used.

src/attester.c
/* SPDX-License-Identifier: BSD-3-Clause */
/*****************************************************************************
 * Copyright 2019, Fraunhofer Institute for Secure Information Technology SIT.
 * All rights reserved.
 ****************************************************************************/

/**
 * @file attester.c
 * @author Michael Eckel (michael.eckel@sit.fraunhofer.de)
 * @author Markus Horn (markus.horn@sit.fraunhofer.de)
 * @brief
 * @version 0.1
 * @date 2019-09-19
 *
 * @copyright Copyright 2019, Fraunhofer Institute for Secure Information
 * Technology SIT. All rights reserved.
 *
 * @license BSD 3-Clause "New" or "Revised" License (SPDX-License-Identifier:
 * BSD-3-Clause).
 */

#include <arpa/inet.h>
#include <coap3/coap.h>
#include <signal.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <tinydtls/session.h>
#include <tss2/tss2_mu.h>
#include <tss2/tss2_tctildr.h>
#include <tss2/tss2_tpm2_types.h>

#include "common/charra_error.h"
#include "common/charra_log.h"
#include "common/charra_macro.h"
#include "core/charra_helper.h"
#include "core/charra_key_mgr.h"
#include "core/charra_tap/charra_tap_cbor.h"
#include "core/charra_tap/charra_tap_dto.h"
#include "util/coap_util.h"
#include "util/config/config_attester_util.h"
#include "util/io_util.h"
#include "util/parser_util.h"
#include "util/tpm2_util.h"

#define CHARRA_UNUSED __attribute__((unused))

/* --- config ------------------------------------------------------------- */

/* quit signal */
static bool quit = false;

/* logging */
#define LOG_NAME "attester"
// #define LOG_LEVEL_CBOR LOG_DEBUG

/* config */
static config_attester config = {0};
#define CBOR_ENCODER_BUFFER_LENGTH 20480  // 20 KiB should be sufficient
// TODO(any): Allocate memory for CBOR buffer with malloc() as logs can be huge.

/**
 * @brief SIGINT handler: set quit to 1 for graceful termination.
 *
 * @param signum the signal number.
 */
static void handle_sigint(int signum);

static void release_data(
        struct coap_session_t* session CHARRA_UNUSED, void* app_ptr);

static void coap_attest_handler(struct coap_resource_t* resource,
        struct coap_session_t* session, const struct coap_pdu_t* request,
        const struct coap_string_t* query, struct coap_pdu_t* response);

/* --- main --------------------------------------------------------------- */

int main(int argc, char** argv) {
    int result = EXIT_FAILURE;

    /* handle SIGINT */
    signal(SIGINT, handle_sigint);

    /* set log level before parsing CLI to be able to print errors. */
    charra_log_set_level(CHARRA_LOG_INFO);

    /* load config from env-vars, files and cli */
    cli_option_code option_code = load_attester_config(argc, argv, &config);
    if (option_code != cli_option_code_continue) {
        // 1 means help message is displayed, -1 means error
        return (option_code == cli_option_code_stop) ? EXIT_SUCCESS
                                                     : EXIT_FAILURE;
    }

    /* set CHARRA and libcoap log levels */
    charra_log_set_level(config.charra_log_level);
    coap_set_log_level(config.coap_log_level);

    trace_log_attester_config(&config);

    /* set varaibles here such that they are valid in case of an 'goto error' */
    coap_context_t* coap_context = NULL;
    coap_endpoint_t* coap_endpoint = NULL;

    if (config.use_dtls_psk && config.use_dtls_rpk) {
        charra_log_error(
                "[" LOG_NAME "] Configuration enables both DTSL with PSK "
                "and DTSL with PKI. Aborting!");
        goto error;
    }

    if (config.use_dtls_psk || config.use_dtls_rpk) {
        // print TLS version when in debug mode
        coap_show_tls_version(LOG_DEBUG);
    }

    if ((config.use_dtls_psk || config.use_dtls_psk) &&
            !coap_dtls_is_supported()) {
        charra_log_error("[" LOG_NAME "] CoAP does not support DTLS but the "
                         "configuration enables DTLS. Aborting!");
        goto error;
    }

    charra_log_info("[" LOG_NAME "] Initializing CoAP in block-wise mode.");
    if ((coap_context = charra_coap_new_context(true)) == NULL) {
        charra_log_error("[" LOG_NAME "] Cannot create CoAP context.");
        goto error;
    }

    if (config.use_dtls_psk) {
        charra_log_info(
                "[" LOG_NAME "] Creating CoAP server endpoint using DTLS-PSK.");
        if (!coap_context_set_psk(coap_context, config.dtls_psk_hint,
                    (uint8_t*)config.dtls_psk_key,
                    strlen(config.dtls_psk_key))) {
            charra_log_error("[" LOG_NAME
                             "] Error while configuring CoAP to use DTLS-PSK.");
            goto error;
        }

        if ((coap_endpoint = charra_coap_new_endpoint(coap_context,
                     config.listen_ip, config.port, COAP_PROTO_DTLS)) == NULL) {
            charra_log_error("[" LOG_NAME "] Cannot create CoAP server "
                             "endpoint based on DTLS-PSK.\n");
            goto error;
        }
    } else if (config.use_dtls_rpk) {
        charra_log_info(
                "[" LOG_NAME "] Creating CoAP server endpoint using DTLS-RPK.");
        coap_dtls_pki_t dtls_pki = {0};

        CHARRA_RC rc = charra_coap_setup_dtls_pki_for_rpk(&dtls_pki,
                config.dtls_rpk_private_key_path,
                config.dtls_rpk_public_key_path,
                config.dtls_rpk_peer_public_key_path,
                config.dtls_rpk_verify_peer_public_key);
        if (rc != CHARRA_RC_SUCCESS) {
            charra_log_error("[" LOG_NAME
                             "] Error while setting up DTLS-RPK structure.");
            goto error;
        }

        if (!coap_context_set_pki(coap_context, &dtls_pki)) {
            charra_log_error("[" LOG_NAME
                             "] Error while configuring CoAP to use DTLS-RPK.");
            goto error;
        }

        if ((coap_endpoint = charra_coap_new_endpoint(coap_context,
                     config.listen_ip, config.port, COAP_PROTO_DTLS)) == NULL) {
            charra_log_error("[" LOG_NAME "] Cannot create CoAP server "
                             "endpoint based on DTLS-RPK.\n");
            goto error;
        }
    } else {
        charra_log_info(
                "[" LOG_NAME "] Creating CoAP server endpoint using UDP.");
        if ((coap_endpoint = charra_coap_new_endpoint(coap_context,
                     config.listen_ip, config.port, COAP_PROTO_UDP)) == NULL) {
            charra_log_error(
                    "[" LOG_NAME
                    "] Cannot create CoAP server endpoint based on UDP.\n");
            goto error;
        }
    }

    /* register CoAP resource and resource handler */
    charra_log_info("[" LOG_NAME "] Registering CoAP resources.");
    charra_coap_add_resource(
            coap_context, COAP_REQUEST_FETCH, "attest", coap_attest_handler);

    /* enter main loop */
    charra_log_debug("[" LOG_NAME "] Entering main loop.");
    while (!quit) {
        /* process CoAP I/O */
        if (coap_io_process(coap_context, COAP_IO_WAIT) == -1) {
            charra_log_error(
                    "[" LOG_NAME "] Error during CoAP I/O processing.");
            goto error;
        }
    }

    result = EXIT_SUCCESS;
    goto finish;

error:
    result = EXIT_FAILURE;

finish:
    /* free CoAP memory */
    charra_free_and_null_ex(coap_endpoint, coap_free_endpoint);
    charra_free_and_null_ex(coap_context, coap_free_context);
    coap_cleanup();

    return result;
}

/* --- function definitions ----------------------------------------------- */

static void handle_sigint(int signum CHARRA_UNUSED) { quit = true; }

static void release_data(
        struct coap_session_t* session CHARRA_UNUSED, void* app_ptr) {
    charra_free_and_null(app_ptr);
}

static void coap_attest_handler(struct coap_resource_t* resource,
        struct coap_session_t* session, const struct coap_pdu_t* request,
        const struct coap_string_t* query, struct coap_pdu_t* response) {
    CHARRA_RC charra_r = CHARRA_RC_SUCCESS;
    int coap_r = 0;
    TSS2_RC tss_r = 0;
    ESYS_TR sig_key_handle = ESYS_TR_NONE;
    /* --- receive incoming data --- */

    charra_log_info(
            "[" LOG_NAME "] Resource '%s': Received message.", "attest");
    coap_show_pdu(LOG_DEBUG, request);

    /* get data */
    size_t data_len = 0;
    const uint8_t* data = NULL;
    size_t data_offset = 0;
    size_t data_total_len = 0;
    if ((coap_r = coap_get_data_large(request, &data_len, &data, &data_offset,
                 &data_total_len)) == 0) {
        charra_log_error("[" LOG_NAME "] Could not get CoAP PDU data.");
        goto error;
    } else {
        charra_log_info(
                "[" LOG_NAME "] Received data of length %zu.", data_len);
        charra_log_info("[" LOG_NAME "] Received data of total length %zu.",
                data_total_len);
    }

    /* unmarshal data */
    charra_log_info("[" LOG_NAME "] Parsing received CBOR data.");
    charra_tap_msg_attestation_request_dto req = {0};
    if ((charra_r = charra_tap_unmarshal_attestation_request(
                 data_len, data, &req)) != CHARRA_RC_SUCCESS) {
        charra_log_error("[" LOG_NAME "] Could not parse CBOR data.");
        goto error;
    }

    /* --- TPM quote --- */

    charra_log_info("[" LOG_NAME "] Preparing TPM quote data.");

    /* nonce */
    if (req.nonce_len > sizeof(TPMU_HA)) {
        charra_log_error("[" LOG_NAME "] Nonce too long.");
        goto error;
    }
    TPM2B_DATA qualifying_data = {.size = 0, .buffer = {0}};
    qualifying_data.size = req.nonce_len;
    memcpy(qualifying_data.buffer, req.nonce, req.nonce_len);

    charra_log_info("[" LOG_NAME
                    "] Received qualifying data (nonce) of length %d:",
            req.nonce_len);
    charra_print_hex(CHARRA_LOG_INFO, req.nonce_len, req.nonce,
            "                                              0x", "\n", false);

    /* PCR selection */
    TPML_PCR_SELECTION pcr_selection = {0};
    if ((charra_r = charra_pcr_selections_to_tpm_pcr_selections(
                 req.pcr_selections_len, req.pcr_selections, &pcr_selection)) !=
            CHARRA_RC_SUCCESS) {
        charra_log_error("[" LOG_NAME "] PCR selection conversion error.");
        goto error;
    }

    /* initialize ESAPI */
    ESYS_CONTEXT* esys_ctx = NULL;
    TSS2_TCTI_CONTEXT* tcti_ctx = NULL;
    if ((tss_r = Tss2_TctiLdr_Initialize(getenv("CHARRA_TCTI"), &tcti_ctx)) !=
            TSS2_RC_SUCCESS) {
        charra_log_error("[" LOG_NAME "] Tss2_TctiLdr_Initialize.");
        goto error;
    }
    if ((tss_r = Esys_Initialize(&esys_ctx, tcti_ctx, NULL)) !=
            TSS2_RC_SUCCESS) {
        charra_log_error("[" LOG_NAME "] Esys_Initialize.");
        goto error;
    }

    /* load TPM key */
    charra_log_info("[" LOG_NAME "] Loading TPM (attestation) key.");
    if ((charra_r = charra_load_tpm2_key(esys_ctx, &sig_key_handle, &config)) !=
            CHARRA_RC_SUCCESS) {
        charra_log_error(
                "[" LOG_NAME "] Could not load TPM (attestation) key.");
        goto error;
    }

    /* determine signature scheme */
    TPMT_SIG_SCHEME sig_scheme = {0};
    if ((tss_r = tpm2_determine_signature_scheme(esys_ctx, sig_key_handle,
                 config.hash_algorithm, config.signature_scheme,
                 &sig_scheme)) != TSS2_RC_SUCCESS) {
        charra_log_error("[" LOG_NAME
                         "] Could not determine signature scheme. Error: %d",
                tss_r);
        goto error;
    }

    /* perform TPM quote */
    charra_log_info("[" LOG_NAME "] Perform TPM2 Quote.");
    TPM2B_ATTEST* attest_buf = NULL;
    TPMT_SIGNATURE* signature = NULL;
    if ((tss_r = tpm2_quote(esys_ctx, sig_key_handle, &pcr_selection,
                 &qualifying_data, &sig_scheme, &attest_buf, &signature)) !=
            TSS2_RC_SUCCESS) {
        charra_log_error(
                "[" LOG_NAME "] TPM2 quote unsuccessful. Error: %d", tss_r);
        goto error;
    } else {
        charra_log_info("[" LOG_NAME "] TPM2 Quote successful.");
    }

    /* --- send response data --- */

    pcr_log_response_dto* pcr_log_responses = NULL;
    pcr_log_responses = malloc(req.pcr_log_len * sizeof(pcr_log_response_dto));

    /* parse log files if requested */
    for (uint32_t i = 0; i < req.pcr_log_len; i++) {
        parse_pcr_log_request(LOG_NAME, config.ima_log_path,
                config.tcg_boot_log_path, req.pcr_logs + i,
                pcr_log_responses + i);
    }

    /* prepare response */
    charra_log_info("[" LOG_NAME "] Preparing response.");

    /* prepare response DTO */
    charra_tap_msg_attestation_response_dto res = {
            .tpm2_quote =
                    {
                            .attestation_data_len = attest_buf->size,
                            .attestation_data =
                                    {0},  // must be memcpy'd, see below
                            .tpm2_signature_len = sizeof(*signature),
                            .tpm2_signature =
                                    {0},  // must be memcpy'd, see below
                    },
            .pcr_log_len = req.pcr_log_len,
            .pcr_logs = pcr_log_responses,
    };
    memcpy(res.tpm2_quote.attestation_data, attest_buf->attestationData,
            res.tpm2_quote.attestation_data_len);
    memcpy(res.tpm2_quote.tpm2_signature, signature,
            res.tpm2_quote.tpm2_signature_len);

    /* clean up */
    charra_free_and_null(signature);
    charra_free_and_null(attest_buf);

    /* marshal response */
    charra_log_info("[" LOG_NAME "] Marshaling response to CBOR.");
    uint32_t res_buf_len = 0;
    uint8_t* res_buf = NULL;
    if ((charra_r = charra_tap_marshal_attestation_response(
                 &res, &res_buf_len, &res_buf)) != CHARRA_RC_SUCCESS) {
        charra_log_error("[" LOG_NAME "] Error marshaling data.");
        goto error;
    }
    charra_log_info("[" LOG_NAME "] Size of marshaled response is %d bytes.",
            res_buf_len);

    // TODO(any) In case an error occurred, an error respone should be sent.
    // TODO(any): The verifier should be able to handle this error reponse.

    /* add response data to outgoing PDU and send it */
    charra_log_info(
            "[" LOG_NAME
            "] Adding marshaled data to CoAP response PDU and send it.");
    coap_pdu_set_code(response, COAP_RESPONSE_CODE_CONTENT);
    if ((coap_r = coap_add_data_large_response(resource, session, request,
                 response, query, COAP_MEDIATYPE_APPLICATION_CBOR, -1, 0,
                 res_buf_len, res_buf, release_data, res_buf)) == 0) {
        charra_log_error("[" LOG_NAME
                         "] Error invoking coap_add_data_large_response().");
        goto error;
    }

error:
    /* free heap objects */
    charra_free_if_not_null(signature);
    charra_free_if_not_null(attest_buf);
    for (uint32_t i = 0; i < req.pcr_log_len; i++) {
        charra_free_if_not_null(pcr_log_responses[i].content);
    }
    charra_free_and_null(req.pcr_logs);
    charra_free_if_not_null(pcr_log_responses);
    // charra_io_free_continuous_file_buffer(&ima_event_log);

    /* flush handles */
    if (sig_key_handle != ESYS_TR_NONE) {
        TSS2_RC r = TSS2_RC_SUCCESS;
        if (config.attestation_key_format ==
                ATTESTER_ATTESTATION_KEY_FORMAT_HANDLE) {
            /* close persistent key */
            r = Esys_TR_Close(esys_ctx, &sig_key_handle);
        } else {
            /* flush transient key */
            r = Esys_FlushContext(esys_ctx, sig_key_handle);
        }
        if (r != TSS2_RC_SUCCESS) {
            charra_log_error(
                    "[" LOG_NAME "] TSS cleanup sig_key_handle failed.");
        }
    }

    /* finalize ESAPI */
    if (esys_ctx != NULL) {
        Esys_Finalize(&esys_ctx);
    }
    if (tcti_ctx != NULL) {
        Tss2_TctiLdr_Finalize(&tcti_ctx);
    }
}
